src/modules/base.ts

```diff
@@ -1,119 +1,138 @@
+
+interface ListenerOptions {
+    /** 
+     * Set to true to only invoke this listener once; it will be removed after first invocation.
+     * false by default.  null/undefined will use default.
+     */
+    once: boolean;
+    /**
+     * Provide a function to further filter the invocation of the listener;
+     * The listener will be called if this method returns a truthy value. null by default.  undefined will also use default.
+     */
+    msgPayloadFilter?: MessagePayloadFilter | null;
+}
+
+interface ListenerConfig {
+    listener: MessageListener;
+    options: ListenerOptions;
+}
 
 /**
  * Base Class for PureCloud Client App APIs
  *
  * @since 1.0.0
  */
 class BaseApi {
+    private _targetPcOrigin: string;
+    private _protocolDetails: ProtocolDetails;
+
+    // ----- Message Listening
+    private _msgListenerCfgs: Record<string, ListenerConfig[]> = {};
+    private _msgHandler = this._onMsg.bind(this);
+
+    /**
+     * Injection point for tests.  Should not be used by api users or extenders.
+     */
+    private _commsUtils = commsUtils;
+
+    /**
+     * Injection point for tests.  Should not be used by api users or extenders.
+     */
+    private _myWindow = window as Window;
+
+    /**
+     * Injection point for tests.  Should not be used by api users or extenders.
+     */
+    private _myParent = (window ? window.parent : undefined);
+
     /**
      * Instantiates the BaseApi
      *
-     * @param {object=} cfg Optional configuration
-     * @param {string=} cfg.targetPcOrigin The origin (protocol, hostname, and port) of the target PC environment (e.g. https://apps.mypurecloud.com). Default is '*'.
-     * @param {string=} cfg.protocol The name of the message protocol under which the message will be sent. Default is purecloud-client-apps.
-     * @param {string=} cfg.protocolAgentName The name of the agent from which the message will be sent. Default is purecloud-client-app-sdk (name of the package).
-     * @param {string=} cfg.protocolAgentVersion The version of the agent from which the message will be sent. Default is the version of the package.
+     * @param cfg Optional configuration
      *
      * @since 1.0.0
      */
-    constructor(cfg = {}) {
+    constructor(cfg: {
+        /** The origin (protocol, hostname, and port) of the target PC environment (e.g. https://apps.mypurecloud.com). Default is '*'. */
+        targetPcOrigin?: string | null,
+        /** The name of the message protocol under which the message will be sent. Default is purecloud-client-apps. */
+        protocol?: string | null;
+        /** The name of the agent from which the message will be sent. Default is purecloud-client-app-sdk (name of the package). */
+        protocolAgentName?: string | null;
+        /** The version of the agent from which the message will be sent. Default is the version of the package. */
+        protocolAgentVersion?: string | null;
+    } = {} as any) {
```

Comment:
The `as any` cast here looks unnecessary?

Reply:
Nice catch 👍